service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/workspace/CloneAllFoldersStep.java

```diff
@@ -0,0 +1,94 @@
+package bio.terra.workspace.service.resource.controlled.flight.clone.workspace;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.common.utils.FlightUtils;
+import bio.terra.workspace.db.FolderDao;
+import bio.terra.workspace.service.folder.model.Folder;
+import bio.terra.workspace.service.job.JobMapKeys;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.FolderKeys;
+import bio.terra.workspace.service.workspace.model.Workspace;
+import com.google.common.collect.ImmutableList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CloneAllFoldersStep implements Step {
+
+  private static final Logger logger = LoggerFactory.getLogger(FindEnabledApplicationsStep.class);
+  private final FolderDao folderDao;
+
+  public CloneAllFoldersStep(FolderDao folderDao) {
+    this.folderDao = folderDao;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    FlightUtils.validateRequiredEntries(
+        context.getInputParameters(), ControlledResourceKeys.SOURCE_WORKSPACE_ID);
+    var sourceWorkspaceId =
+        context.getInputParameters().get(ControlledResourceKeys.SOURCE_WORKSPACE_ID, UUID.class);
+    var destinationWorkspaceId =
+        context
+            .getInputParameters()
+            .get(JobMapKeys.REQUEST.getKeyName(), Workspace.class)
+            .getWorkspaceId();
+
+    // Create and clone all folders
+    ImmutableList<Folder> foldersResult = folderDao.listFoldersInWorkspace(sourceWorkspaceId);
+    // Use Map<String, String> rather than Map<UUID, UUID> to avoid JSON deserialization error
+    Map<String, String> folderIdMap = new HashMap<>();
+    if (foldersResult != null) {
+      for (Folder folder : foldersResult) {
+        UUID destinationFolderId = UUID.randomUUID();
```

Comment:
This PR

- Creates a new UUID for destinationFolderId (UUID.randomUUID())
- Stores folder id in FolderKeys.FOLDER_IDS_TO_CLONE_MAP
- FindResourcesToCloneStep reads FolderKeys.FOLDER_IDS_TO_CLONE_MAP

Some background: When you close a resource, [properties are cloned.](https://cs.github.com/DataBiosphere/terra-workspace-manager/blob/7f9cd647850acb6786f57e1e7a20b4af3fe57e86/service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/workspace/WorkspaceCloneUtils.java#L382) Here we [intentionally don't clone folder id property.](https://cs.github.com/DataBiosphere/terra-workspace-manager/blob/7f9cd647850acb6786f57e1e7a20b4af3fe57e86/service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/workspace/WorkspaceCloneUtils.java#L408)

We did that because if you clone resource from workspace A to workspace B, the folder id in workspace A won't exist in workspace B.

But in this case, the folder id *will* exist. I think it's cleaner to:

- Instead of creating a new UUID for destinationFolderId, use sourceFolderId
- Delete FolderKeys.FOLDER_IDS_TO_CLONE_MAP. You can also delete ControlledResourceKeys.DESTINATION_FOLDER_ID, since you can use the source folder id. A bunch of files can be reverted.
- Add boolean parameter `shouldCloneFolderIdProperty` property to `buildDestinationReferencedResource`. If cloning an individual resource, `shouldCloneFolderIdProperty` will be false. If cloning workspace, `shouldCloneFolderIdProperty` will be true.

WDYT?